test/ammo.mjs

```diff
@@ -0,0 +1,1633 @@
+/**
+ * Get all types that descend from a searched type.
+ *
+ * @param {string} type - Searched type name.
+ * @param {boolean} [classOnly] - Whether to return only types that have constructor.
+ * @returns {string[]} List of type names where type is descendant of searched type.
+ */
+function allOfAmmoType(type, classOnly = false) {
+    return Object.keys(AmmoTypes).filter(t => isOfAmmoType(AmmoTypes[t], type) && (classOnly ? typeof Ammo[t] === 'function' : true));
+}
+
+/**
+ * Creates a new Ammo type for testing.
+ *
+ * @param {string} name - The name of this type.
+ * @param {object} definition - The definition for this type.
+ * @param {object} [parentName] - The parent type name.
+ */
+function createType(name, definition, parentName) {
+    const obj = parentName ? {
+        ...AmmoTypes[parentName],
+        ...definition
+    } : definition;
+
+    if (obj.__type) {
+        obj.__type.push(name);
+    } else {
+        obj.__type = [name];
+    }
+
+    AmmoTypes[name] = obj;
+
+    if (obj._constructor) {
+        const fn = function (...args) {
+            obj._constructor(...args);
+            return obj;
+        };
+
+        fn._obj = obj;
+
+        Ammo[name] = fn;
+    } else {
+        Ammo[name] = obj;
+    }
+}
+
+// Linear Math
+
+createType('btVector3', {
+    _constructor: (x, y, z) => {
+        if (x !== undefined) {
+            assertNum(x);
+            assertNum(y);
+            assertNum(z);
+        }
+    },
+
+    length: () => 0,
+    x: () => 0,
+    y: () => 0,
+    z: () => 0,
+
+    setX: x => assertNum(x),
+    setY: y => assertNum(y),
+    setZ: z => assertNum(z),
+
+    setValue: (x, y, z) => {
+        assertNum(x);
+        assertNum(y);
+        assertNum(z);
+    },
+
+    normalize: function () { },
+    dot: v => assertAmmoType(v, 'btVector3'),
+
+    op_mul: x => assertNum(x),
+    op_add: v => assertAmmoType(v, 'btVector3'),
+    op_sub: v => assertAmmoType(v, 'btVector3')
+});
+
+createType('btVector4', {
+    _constructor: (x, y, z, w) => {
+        if (x !== undefined) {
+            assertNum(x);
+            assertNum(y);
+            assertNum(z);
+            assertNum(w);
+        }
+    },
+
+    w: () => 0,
+
+    setValue: (x, y, z, w) => {
+        assertNum(x);
+        assertNum(y);
+        assertNum(z);
+        assertNum(w);
+    },
+
+    dot: v => assertAmmoType(v, 'btVector4'),
+
+    op_mul: x => assertNum(x),
+    op_add: v => assertAmmoType(v, 'btVector4'),
+    op_sub: v => assertAmmoType(v, 'btVector4')
+}, 'btVector3');
+
+createType('btQuadWord', {
+    x: () => 0,
+    y: () => 0,
+    z: () => 0,
+    w: () => 0,
+
+    setX: x => assertNum(x),
+    setY: y => assertNum(y),
+    setZ: z => assertNum(z),
+    setW: w => assertNum(w)
+});
+
+createType('btQuaternion', {
+    _constructor: (x, y, z, w) => {
+        if (x !== undefined) {
+            assertNum(x);
+            assertNum(y);
+            assertNum(z);
+            assertNum(w);
+        }
+    },
```

Comment:
I know we do call `new Ammo.btQuaternion()` without arguments, but technically this is just wrong if you take a look at the actual WebIDL:

https://github.com/kripken/ammo.js/blob/1ed8b58c7058a5f697f2642ceef8ee20fdd55e10/ammo.idl#L62-L64

So in a way if there are actual errors, this won't find it, because you just assume that as things are right now is the definition of "correct", even though there may be bugs?

Sorry for stupid question, but I just still don't understand what this really does or how this helps to find bugs in Ammo?

Reply:
This doesn't suppose to find bugs in Ammo. Ammo handles its own tests, we don't care about it. This can catch bugs, like the one we had recently, where we were adding a body to dynamics world twice, or an earlier one, where we were destroying the same body twice, etc. This allows to make sure the engine calls correct Ammo methods, correct amount of times and in correct order. What happens inside Ammo is of no importance, thus we stub those.

Reply:
About quaternion - we should init those with zeros, as per API. If we don't, those are initialized with garbage, so unless we immediately set values for all components right after, it can introduce hard to debug bugs.